server/src/main/java/org/elasticsearch/search/query/TopDocsCollectorContext.java

```diff
@@ -357,6 +364,22 @@ static int shortcutTotalHitCount(IndexReader reader, Query query) throws IOExcep
                 count += context.reader().docFreq(term);
             }
             return count;
+        } else if (query.getClass() == DocValuesFieldExistsQuery.class && reader.hasDeletions() == false) {
+            final String field = ((DocValuesFieldExistsQuery) query).getField();
+            int count = 0;
+            for (LeafReaderContext context : reader.leaves()) {
+                FieldInfos fieldInfos = context.reader().getFieldInfos();
+                FieldInfo fieldInfo = fieldInfos.fieldInfo(field);
+                if (fieldInfo.getPointIndexDimensionCount() > 0) {
+                    count += PointValues.getDocCount(context.reader(), field);
+                } else if (fieldInfo.getIndexOptions() != IndexOptions.NONE) {
+                    Terms terms = context.reader().terms(field);
+                    count += terms.getDocCount();
```

Comment:
I think you need to check that `terms` is not null here or you might have errors if the term exists on none of the documents in the segment.

Reply:
Is there a simple way to provoke this in edge case in tests? If not I'll just add the null check regardless.

Reply:
Just realized the tests just ran into something similar in the points branch (I think). Its reproducible, will see if I can recreate that in tests as well.

Reply:
It looks like the RandomIndexWriter takes care of running into these cases occasionally, The NPE was triggered in about 1 of 50 runs.

Reply:
You can trigger it explicitly by creating a first segment that only consists of an empty document. Assuming this segment is not merged away, it should cause the NPE.